src/aspire/source/__init__.py

```diff
@@ -236,52 +324,56 @@ def cache(self, im=None):
             im = self.images()
         self._im = im
 
-    def images(self, start=0, num=np.inf, indices=None, *args, **kwargs):
+    def images(self, start=0, num=np.inf, indices=None, batch_size=256, *args, **kwargs):
```

Comment:
Ideally, batching should be done at the next level up. This is the reason for providing the `(start, num)` indices to `images()`.

Reply:
So the reason I added this as an argument:

In the new world where images are run through a `Pipeline` (where one has downsampled the source down the line and is thus confident that (in the extreme case) one can call the `images()` method to get *all* images, without running out of memory), there is still a distinction between how many images the caller wants (in this case *all* of them), vs. how the pipeline processing should be batched. The `batch_size` specifies how to reasonably run the pipeline.

In other words, when saying `src.downsample(8)` followed by `im = src.images()` (which should both work properly when used in conjunction), a `batch_size` is still implied. We can call this something else to make it clear. This will be a non-issue if we eventually end up taking out the whole `Pipeline` out of `images()`.

Reply:
Ok I see. What I'm worried about here is that we might end up getting two levels of batching (one by the caller of `images()` and one inside of `images()` itself). This would be less efficient than a single level.

The use case of calling `images()` to get *all* the images is quite restrictive, and should ideally only be done during prototyping and testing. Since any general-purpose code *cannot* rely on all the images fitting into memory, it would be written to batch its input as needed. This issue would mostly come up in interactive work or during basic tests (where it should be easy to identify and fix).

A compromise may be to increase the default batch size here (inside `images()`) to something relatively large (say `batch_size=32678`)? That way, it is less likely to affect smaller calls (which may take `n=8192` or comparable).

Reply:
Default `batch_size`  increased to `32768`, with the justification in docstrings.